examples/hif2a/fit_to_multiple_rbfes.py

```diff
@@ -0,0 +1,431 @@
+from pathlib import Path
+
+# locations relative to project root
+root = Path(__file__).parent.parent.parent
+path_to_protein = str(root.joinpath('tests/data/hif2a_nowater_min.pdb'))
+path_to_ff = str(root.joinpath('ff/params/smirnoff_1_1_0_ccc.py'))
+
+# locations relative to example folder
+path_to_results = Path(__file__).parent
+path_to_transformations = str(path_to_results.joinpath('relative_transformations.pkl'))
+
+# load and construct forcefield
+with open(path_to_ff) as f:
+    ff_handlers = deserialize_handlers(f.read())
+
+forcefield = Forcefield(ff_handlers)
+
+
+def wrap_method(args, fxn):
+    return fxn(*args)
+
+
+def construct_lambda_schedule(num_windows):
+    """manually optimized by YTZ
+
+    TODO: should move this into a common module, probably in free_energy
+    """
+
+    A = int(.35 * num_windows)
+    B = int(.30 * num_windows)
+    C = num_windows - A - B
+
+    # Empirically, we see the largest variance in std <du/dl> near the endpoints in the nonbonded
+    # terms. Bonded terms are roughly linear. So we add more lambda windows at the endpoint to
+    # help improve convergence.
+    lambda_schedule = np.concatenate([
+        np.linspace(0.0, 0.25, A, endpoint=False),
+        np.linspace(0.25, 0.75, B, endpoint=False),
+        np.linspace(0.75, 1.0, C, endpoint=True)
+    ])
+
+    assert len(lambda_schedule) == num_windows
+
+    return lambda_schedule
+
+
+def type_check_handlers(handlers):
+    """check that handlers for charges and vdW parameters are compatible with those in forcefield"""
+    for handle_type in handlers:
+        if handle_type == nonbonded.AM1CCCHandler:
+            # sanity check as we have other charge methods that exist
+            assert handle_type == type(forcefield.q_handle)
+
+        elif handle_type == nonbonded.LennardJonesHandler:
+            # sanity check again, even though we don't have other lj methods currently
+            assert handle_type == type(forcefield.lj_handle)
+
+
+def _print_result(lamb, result):
+    """
+    TODO: include type hint here for ambiguous argument name "result"
+    TODO: add units
+    TODO: move message into a __str__ method for a result object?
+    TODO: replace print with logger
+    """
+    bonded_du_dl, nonbonded_du_dl, grads_and_handles = result
+
+    unit = "kJ/mol"
+
+    message = f"""
+    lambda {lamb:.3f}
+    bonded dU/dlambda (in {unit})
+        mean = {np.mean(bonded_du_dl):.3f}
+        stddev = {np.std(bonded_du_dl):.3f}
+    nonbonded dU/dlambda (in {unit})
+        mean = {np.mean(nonbonded_du_dl):.3f}
+        stddev = {np.std(nonbonded_du_dl):.3f}
+    """
+    print(message)
+
+
+def _mean_du_dlambda(result):
+    """summarize result of rfe.host_edge into mean du/dl
+
+    TODO: refactor where this analysis step occurs
+    """
+    bonded_du_dl, nonbonded_du_dl, _ = result
+    return np.mean(bonded_du_dl + nonbonded_du_dl)
+
+
+def _update_combined_handle_and_grads(ghs, combined_handle_and_grads):
```

Comment:
Looking back, I wish there was a slightly more elegant way to compute these derivatives. One possible solution is to just treat all of this the jax/AD way, i.e. define a custom vjp function on observables of system parameters.

Let ```p``` be forcefield parameters, and ```q(p)``` be system parameters derived from ```p```. Let ```G(q)``` be our free energy estimate. Let ```L(G)``` be our loss function defined relative to some label.

Given:
```
L(G(q(p)))
```
Its derivative is
```
dL/dG . dG/dq . dq/dp
```
We can use jax's vjp system to directly compute outer ```dL/dG``` and ```dq/dp```. However, the inner ```dG/dq``` is done currently using a custom_op of sorts. In theory we can declare ```G(q)``` a primitive and augment it with a defvjp so that the chain rule can be done automatically.

Reply:
PS - definitely don't try and actually do this right now. We should save it for another PR after extensive offline discussions.

Reply:
We have discussed this a bit offline, and @proteneer has implemented this approach in https://github.com/proteneer/timemachine/pull/330 ! I think this represents an immediate improvement, reducing the amount that needs to be managed explicitly in application scripts like the ones developed in the current PR. It also improves our flexibility in prototyping loss functions (and other differentiable computations that accept a free energy estimate as an argument). We should probably merge this tomorrow as far as I'm concerned!

However, I do have a reservation about the approach I would like to bookmark for when we possibly revisit this design down the road.

I think being able to iterate quickly on design choices made in the free-energy-estimation layer represents a large source of overall performance gains, and I worry that this approach might hinder design changes at this layer, to the extent that it "locks in" a specific free-energy-estimation protocol.

The approach I would favor in the long run is to provide accelerated wrappers for the essential, expensive parts of any free energy computation that are deterministic and amenable to autodiff -- keeping any experimental / noisy / biased approximations in prototyping-friendly functions that call these (rather than pushing any of these approximations into custom chain rule code). Specifically, I would want to make sure we have something like "`U(x, lam, params)`, registering the handlers required for `grad(U, argnums=[0,1,2])`, `value_and_grad(U, argnums=[0,1,2])`, and perhaps `vmap(U, argnums=[0, 1])`". Most differentiable free energy estimators I can think of only require to interact with these, and can be implemented in prototyping-friendly (autodiff-able!) Python / Jax in terms of `U` and its derivatives (possibly in accelerator-friendly batches via `vmap`).

Whenever we find that having the free energy estimators written in prototyping-friendly Python / Jax on top of this stable core of functions is too slow, then I think there would be clearer benefit of pushing more of the current best-performing free-energy-estimation pipeline into a primitive function like `G(q)` and accompanying it with custom chain rule code.

Reply:
Some notes on the speed/perf side of things since I've tried to make it fast with pure jax:

1) Nonbonded ```U``` and their derivatives will require a neighborlist for any realistic system with > 1000 atoms.
2) The current reference implementations do not implement a proper neighborlist, since writing a performant neighborlist implemention via jax/numpy is **hard**. We're literally talking about 100x slower. 
3) There's not much to be gained with ```vmap```, since the system sizes are practically so large that a single calculation will probably saturate the GPU.

However, doing a batched variant of the potential functions via a custom_vjp is not difficult, and you could more or less amortize out the cost of the memory allocation.

Reply:
On the more general side of things:

It would not be hard to modify the current ```deltaG``` estimator from one that computes ```xs``` and ```dU/dx, dU/dparams, dU/dlambda``` on the fly to one that takes in a pre-computed ```xs``` and post-processes the ```dU/dparams```. We just need to be cognizant of the GPU RAM requirements when computing ```dU/dparams``` for each ```x``` in ```xs```.

Reply:
Sorry for my unclarity! I didn't mean to suggest reverting to `U(x, lam, params)` or its derivatives in pure Jax (which would incur a significant slow-down for the reasons you mention here, and that you've mentioned before!). I was suggesting to wrap the accelerated implementations of` U` and its derivatives you've already implemented, in a way that I think would allow for simpler access in other Jax application code that calls `U` or its derivatives.

For example, if we were implementing the forward EXP estimator of a free energy difference, I would want the calls to `U(xs, 0.0, params)` and `U(xs, 1.0, params)` to be accelerated in timemachine, but let Jax autodiff machinery handle any subsequent chain-rule stuff involved in actually computing the EXP estimate from the potential energy differences, and in any downstream computations. The approach I thought you were proposing, if applied to this example, would be to implement a Jax primitive `G(params)` that runs a simulation at lam=0.0, performs the reweighting from lam=0.0 to lam=1.0, and implements the chain-rule for logsumexp, all as a single unit.